src/scheduler/pct.rs

```diff
@@ -90,13 +104,19 @@ impl Scheduler for PctScheduler {
     }
 
     fn next_task(&mut self, runnable: &[TaskId], current: Option<TaskId>, is_yielding: bool) -> Option<TaskId> {
-        // If any new tasks were created, assign them priorities at random
-        let known_tasks = self.priority_queue.len();
-        let max_tasks = usize::from(*runnable.iter().max().unwrap());
-
-        for tid in known_tasks..1 + max_tasks {
-            let index = self.rng.gen_range(0, self.priority_queue.len() + 1);
-            self.priority_queue.insert(index, TaskId::from(tid));
+        // If any new tasks were created, assign them priorities by randomly swapping them with an
+        // existing task's priority, so we maintain the invariant that every priority is distinct
+        let max_known_task = self.priorities.len();
+        let max_new_task = usize::from(*runnable.iter().max().unwrap());
+        for new_task_id in max_known_task..1 + max_new_task {
+            let target_task_id = TaskId::from(self.rng.gen_range(0, self.priorities.len()));
```

Comment:
This code ensures that a new task always gets a priority lower than some existing task.  (Note that previously we chose a random number in the range 0..self.priority_queue.len()+1 so a new task could go to the end.)  Maybe generate a target_task_id in range 0..self.priorities.len()+1?

Reply:
Yeah, nice catch, done.